src/box/sql/vdbe.c

```diff
@@ -467,12 +467,110 @@ int sqlVdbeExec(Vdbe *p)
 	}
 #endif
 
+#if defined(SQL_DEBUG)
+
+#define VDBE_TRACE() \
+do{ \
+if ((p->sql_flags & SQL_VdbeTrace) != 0) { \
+    u8 opProperty = sqlOpcodeProperty[pOrigOp->opcode]; \
+    if (rc!=0) printf("rc=%d\n",rc); \
+    if (opProperty & (OPFLG_OUT2)) { \
+	registerTrace(pOrigOp->p2, &aMem[pOrigOp->p2]); \
+    } \
+    if (opProperty & OPFLG_OUT3) { \
+        registerTrace(pOrigOp->p3, &aMem[pOrigOp->p3]); \
+    } \
+} \
+}while(0)
+
+#define CHECK_OPCODE() \
+do{ \
+assert(rc == 0); \
+assert(pOp>=aOp && pOp<&aOp[p->nOp]); \
+nVmStep++; \
+if ((p->sql_flags & SQL_VdbeTrace) != 0) \
+    sqlVdbePrintOp(stdout, (int)(pOp - aOp), pOp); \
+{ \
+    u8 opProperty = sqlOpcodeProperty[pOp->opcode]; \
+    if ((opProperty & OPFLG_IN1)!=0) { \
+        assert(pOp->p1>0); \
+	assert(pOp->p1<=(p->nMem+1 - p->nCursor)); \
+	assert(memIsValid(&aMem[pOp->p1])); \
+	assert(sqlVdbeCheckMemInvariants(&aMem[pOp->p1])); \
+	REGISTER_TRACE(p, pOp->p1, &aMem[pOp->p1]); \
+    } \
+    if ((opProperty & OPFLG_IN2)!=0) { \
+        assert(pOp->p2>0); \
+	assert(pOp->p2<=(p->nMem+1 - p->nCursor)); \
+	assert(memIsValid(&aMem[pOp->p2])); \
+	assert(sqlVdbeCheckMemInvariants(&aMem[pOp->p2])); \
+	REGISTER_TRACE(p, pOp->p2, &aMem[pOp->p2]); \
+    } \
+    if ((opProperty & OPFLG_IN3)!=0) { \
+	assert(pOp->p3>0); \
+	assert(pOp->p3<=(p->nMem+1 - p->nCursor)); \
+	assert(memIsValid(&aMem[pOp->p3])); \
+	assert(sqlVdbeCheckMemInvariants(&aMem[pOp->p3])); \
+	REGISTER_TRACE(p, pOp->p3, &aMem[pOp->p3]); \
+    } \
+    if ((opProperty & OPFLG_OUT2)!=0) { \
+	assert(pOp->p2>0); \
+	assert(pOp->p2<=(p->nMem+1 - p->nCursor)); \
+	memAboutToChange(p, &aMem[pOp->p2]); \
+    } \
+    if ((opProperty & OPFLG_OUT3)!=0) { \
+	assert(pOp->p3>0); \
+	assert(pOp->p3<=(p->nMem+1 - p->nCursor)); \
+	memAboutToChange(p, &aMem[pOp->p3]); \
+    } \
+} \
+pOrigOp = pOp; \
+}while(0)
+
+#else
+
+#define VDBE_TRACE() \
+do{ \
+}while(0)
+
+#define CHECK_OPCODE() \
+do{ \
+assert(rc == 0); \
+assert(pOp>=aOp && pOp<&aOp[p->nOp]); \
+nVmStep++; \
+}while(0)
+
+#endif
+
+#if defined(SQL_WITHOUT_COMPUTED_GOTO)
+
+#define EXECUTE(opname) case opname:
+#define NEXT(opcode)    goto start_eval
+#define SWITCH(opcode)  start_eval: switch(opcode)
+
+#else
+
+#define EXECUTE(opname) Exec_##opname:
+#define NEXT(opcode)    goto *dispatch_table[opcode]
+#define SWITCH(opcode)  goto *dispatch_table[opcode];
+
+#endif
+
 #define DISPATCH() \
 do{ \
+VDBE_TRACE(); \
 pOp++; \
-goto *dispatch_table[pOp->opcode]; \
+CHECK_OPCODE(); \
+NEXT(pOp->opcode); \
 }while(0)
 
+#define START_EVAL() \
```

Comment:
This macro seems to be redundant: just inline the code

Reply:
Fixed.